File: cache_interceptor/src/test/java/uk/co/glass_software/android/cache_interceptor/interceptors/cache/DatabaseManagerIntegrationTest.java
package uk.co.glass_software.android.cache_interceptor.interceptors.cache;

import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;

import org.junit.Before;
import org.junit.Test;

import java.util.Date;

import io.reactivex.Observable;
import uk.co.glass_software.android.cache_interceptor.base.BaseIntegrationTest;
import uk.co.glass_software.android.cache_interceptor.base.network.model.TestResponse;
import uk.co.glass_software.android.cache_interceptor.response.ResponseMetadata;
import uk.co.glass_software.android.cache_interceptor.utils.Function;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static uk.co.glass_software.android.cache_interceptor.interceptors.cache.CacheToken.Status.STALE;

@SuppressWarnings("unchecked")
public class DatabaseManagerIntegrationTest extends BaseIntegrationTest {
    
    private TestResponse stubbedResponse;
    private CacheToken<TestResponse> cacheToken;
    private Observable<TestResponse> mockUpstream;
    
    private DatabaseManager target;
    
    @Before
    public void setUp() throws Exception {
        CacheToken requestToken = CacheToken.newRequest(
                TestResponse.class,
                "apiUrl",
                "",
                5
        );
        
        stubbedResponse = assetHelper.getStubbedResponse(TestResponse.STUB_FILE, TestResponse.class).blockingFirst();
        stubbedResponse.setMetadata(ResponseMetadata.create(requestToken, null));
        
        mockUpstream = mock(Observable.class);
        
        Date fetchDate = new Date();
        Date expiryDate = new Date(fetchDate.getTime() + 12345);
        
        cacheToken = CacheToken.caching(requestToken,
                                        mockUpstream,
                                        fetchDate,
                                        fetchDate,
                                        expiryDate
        );
        
        stubbedResponse.getMetadata().setCacheToken(cacheToken);
        
        CacheInterceptorBuilder.Holder holder = new CacheInterceptorBuilder.Holder();
        new CacheInterceptorBuilderHelper().build(application, holder);
        
        target = holder.databaseManager;
    }
    
    @Test
    public void testCacheAndFlush() throws Exception {
        assertNull("Cache should not contain anything", target.getCachedResponse(mockUpstream, cacheToken));
        
        target.cache(stubbedResponse);
        
        TestResponse cachedResponse = target.getCachedResponse(mockUpstream, cacheToken);
        cachedResponse.setMetadata(stubbedResponse.getMetadata()); //ignore metadata, covered by unit test
        assertEquals("Cached response didn't match the original", stubbedResponse, cachedResponse);
        
        target.flushCache();
        
        assertNull("Cache should not contain anything", target.getCachedResponse(mockUpstream, cacheToken));
    }
    
    //used by other tests to preserve encapsulation
    @VisibleForTesting
    public static <E extends Exception & Function<E, Boolean>, R extends ResponseMetadata.Holder<R, E>> void cache(DatabaseManager databaseManager,
                                                                                                                   R response,
                                                                                                                   CacheToken cacheToken) {
        response.setMetadata(ResponseMetadata.create(cacheToken, null));
        databaseManager.cache(response);
    }
    
    //used by other tests to preserve encapsulation
    @VisibleForTesting
    public static <E extends Exception & Function<E, Boolean>, R extends ResponseMetadata.Holder<R, E>> R getCachedResponse(DatabaseManager databaseManager,
                                                                                                                            Observable<R> upstream,
                                                                                                                            CacheToken cacheToken) {
        return databaseManager.getCachedResponse(upstream, cacheToken);
    }
    
    //used by other tests to preserve encapsulation
    @VisibleForTesting
    public static <E extends Exception & Function<E, Boolean>, R extends ResponseMetadata.Holder<R, E>> CacheToken<R> getCachedCacheToken(CacheToken<R> cacheToken,
                                                                                                                                          @NonNull Observable<R> refreshObservable,
                                                                                                                                          @NonNull Date cacheDate,
                                                                                                                                          @NonNull Date expiryDate) {
        return CacheToken.cached(cacheToken,
                                 refreshObservable,
                                 cacheDate,
                                 expiryDate
        );
    }
    //used by other tests to preserve encapsulation
    @VisibleForTesting
    public static void prepareSpyCacheManagerToReturnStale(CacheManager spyCacheManager) {
        doReturn(STALE).when(spyCacheManager).getCachedStatus(any());
    }
}

File: cache_interceptor/src/test/java/uk/co/glass_software/android/dejavu/interceptors/internal/cache/database/DatabaseManagerIntegrationTest.kt
package uk.co.glass_software.android.dejavu.interceptors.internal.cache.database

import org.junit.Test
import uk.co.glass_software.android.dejavu.configuration.CacheInstruction.Operation.Expiring.Cache
import uk.co.glass_software.android.dejavu.configuration.CacheInstruction.Operation.Expiring.Refresh
import uk.co.glass_software.android.dejavu.interceptors.internal.cache.token.CacheStatus
import uk.co.glass_software.android.dejavu.interceptors.internal.error.Glitch
import uk.co.glass_software.android.dejavu.test.BaseIntegrationTest
import uk.co.glass_software.android.dejavu.test.assertNullWithContext
import uk.co.glass_software.android.dejavu.test.instructionToken
import java.util.*

internal class DatabaseManagerIntegrationTest : BaseIntegrationTest<DatabaseManager<Glitch>>({ it.databaseManager() }) {

    @Test
    fun `GIVEN that a response is not cached THEN it should not be returned`() {
        val cachedResponse = target.getCachedResponse(instructionToken(), 500)

        assertNullWithContext(
                cachedResponse,
                "Response should be null"
        )
    }

    @Test
    fun `GIVEN that a response is cached THEN it should be returned`() {
        val stubbedResponse = getStubbedTestResponse()

        target.cache(stubbedResponse, null).blockingAwait()

        val cachedResponse = target.getCachedResponse(instructionToken(), 500)

        assertResponse(
                stubbedResponse,
                cachedResponse,
                CacheStatus.CACHED
        )
    }

    @Test
    fun `GIVEN that a response has expired THEN it should have a STALE status`() {
        val stubbedResponse = getStubbedTestResponse(instructionToken(
                Cache(durationInMillis = -1L)
        ))

        val instructionToken = stubbedResponse.metadata.cacheToken

        target.cache(stubbedResponse, null).blockingAwait()

        val cachedResponse = target.getCachedResponse(instructionToken, 500)

        assertResponse(
                stubbedResponse,
                cachedResponse,
                CacheStatus.STALE,
                expiryDate = Date(NOW.time - 1L)
        )
    }

    @Test
    fun `GIVEN that a response has not expired AND the operation is INVALIDATE THEN it should have a STALE status`() {
        val stubbedResponse = getStubbedTestResponse()
        val instructionToken = stubbedResponse.metadata.cacheToken

        target.cache(stubbedResponse, null).blockingAwait()
        target.invalidate(instructionToken)

        val cachedResponse = target.getCachedResponse(instructionToken, 500)

        assertResponse(
                stubbedResponse,
                cachedResponse,
                CacheStatus.STALE,
                expiryDate = Date(0)
        )
    }

    @Test
    fun `GIVEN that a response has not expired AND the operation is REFRESH THEN it should have a STALE status`() {
        val stubbedResponse = getStubbedTestResponse()

        target.cache(stubbedResponse, null).blockingAwait()

        val cachedResponse = target.getCachedResponse(instructionToken(), 500)

        assertResponse(
                stubbedResponse,
                cachedResponse,
                CacheStatus.CACHED
        )

        val refreshToken = instructionToken(Refresh())
        val refreshResponse = target.getCachedResponse(refreshToken, 500)

        assertResponse(
                getStubbedTestResponse(refreshToken),
                refreshResponse,
                CacheStatus.STALE,
                expiryDate = Date(0)
        )
    }

    @Test
    fun `GIVEN that a response has not expired AND the operation is not REFRESH or INVALIDATE THEN it should have a FRESH status`() {

    }

    @Test
    fun `GIVEN that a response is cached AND the cache is cleared for that type THEN it should not be returned AND the database should contain other types`() {

    }

    @Test
    fun `GIVEN that a response is cached AND the whole cache is cleared for that type THEN it should not be returned AND the database should be empty`() {

    }

    @Test
    fun `GIVEN that a response must be cached encrypted THEN it is returned encrypted`() {

    }

    @Test
    fun `GIVEN that a response must be cached compressed THEN it is returned compressed`() {

    }

    @Test
    fun `GIVEN that a response must be cached compressed and encrypted THEN it is returned compressed and encrypted`() {

    }

    @Test
    fun `GIVEN that a response must be cached encrypted as per previous response THEN it is returned encrypted`() {

    }

    @Test
    fun `GIVEN that a response must be cached compressed as per previous response THEN it is returned compressed`() {

    }

    @Test
    fun `GIVEN that a response must be cached compressed and encrypted as per previous response THEN it is returned compressed and encrypted`() {

    }

    @Test
    fun `GIVEN that a response must be cached as is as per previous response THEN it is returned as is`() {

    }

}